pkg/plugin/resources.go

```diff
@@ -464,6 +466,154 @@ func (a *App) handleLLMState(w http.ResponseWriter, req *http.Request) {
 	}
 }
 
+// provisionedPlugin is the response returned by a call to grafana.com's provisioned-plugin's endpoint
+// for an specific instance's plugin.
+type provisionedPlugin struct {
+	ID int `json:"id"`
+}
+
+// getPluginID gets the *ID* of the *provisioned plugin* from grafana.com.
+// Note that this differs to the plugin ID referred to by the `backend.CallResourceRequest`,
+// which is 'grafana-llm-app'
+func getPluginID(ctx context.Context, slug string, grafanaAppURL string, saToken string, gcomAPIKey string) (int, error) {
+	gcomPath := "/api/gnet/instances/" + slug + "/provisioned-plugins/grafana-llm-app"
+	req, err := http.NewRequestWithContext(ctx, "GET", grafanaAppURL+gcomPath, nil)
+	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", saToken))
+	req.Header.Set("X-Api-Key", gcomAPIKey)
+	if err != nil {
+		return 0, fmt.Errorf("create http request: %w", err)
+	}
+	respBody, err := doRequest(req)
+	if err != nil {
+		return 0, err
+	}
+	plugin := provisionedPlugin{}
+	err = json.Unmarshal(respBody, &plugin)
+	if err != nil {
+		return 0, fmt.Errorf("unmarshal json: %w", err)
+	}
+	return plugin.ID, nil
+}
+
+type pluginSettings struct {
+	JSONData       map[string]interface{} `json:"jsonData"`
+	SecureJSONData map[string]string      `json:"secureJsonData"`
+}
+
+func (a *App) mergeSecureJSONData(body io.ReadCloser) (url.Values, error) {
+	// Read the request body
+	if body == nil {
+		return nil, errors.New("request body required")
+	}
+	b, err := io.ReadAll(body)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read request body to bytes %w", err)
+	}
+	// Unmarshal the request body to JSON
+	var requestData pluginSettings
+	err = json.Unmarshal(b, &requestData)
+	if err != nil {
+		return nil, fmt.Errorf("failed to unmarshal request body to JSON %w", err)
+	}
+
+	// Update mandatory fields
+	requestData.SecureJSONData[encodedTenantAndTokenKey] = base64.StdEncoding.EncodeToString([]byte(a.settings.DecryptedSecureJSONData[encodedTenantAndTokenKey]))
+
+	// Insert existing plugin secureJSONData fields if missing from request
+	for key, value := range a.settings.DecryptedSecureJSONData {
+		if existingValue, exists := requestData.SecureJSONData[key]; !exists || existingValue == "" {
```

Comment:
Should this only check `exists`? Otherwise there would be no way to unset a piece of secure json data by setting it to ""?